core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/stream/bi/ComposeFourBiCollector.java

```diff
@@ -0,0 +1,114 @@
+    private final Function<ResultHolder4_, Result4_> fourthFinisher;
+
+    ComposeFourBiCollector(BiConstraintCollector<A, B, ResultHolder1_, Result1_> first,
+            BiConstraintCollector<A, B, ResultHolder2_, Result2_> second,
+            BiConstraintCollector<A, B, ResultHolder3_, Result3_> third,
+            BiConstraintCollector<A, B, ResultHolder4_, Result4_> fourth,
+            QuadFunction<Result1_, Result2_, Result3_, Result4_, Result_> composeFunction) {
+        this.first = first;
+        this.second = second;
+        this.third = third;
+        this.fourth = fourth;
+        this.composeFunction = composeFunction;
+
+        this.firstSupplier = first.supplier();
+        this.secondSupplier = second.supplier();
+        this.thirdSupplier = third.supplier();
+        this.fourthSupplier = fourth.supplier();
+
+        this.firstAccumulator = first.accumulator();
+        this.secondAccumulator = second.accumulator();
+        this.thirdAccumulator = third.accumulator();
+        this.fourthAccumulator = fourth.accumulator();
+
+        this.firstFinisher = first.finisher();
+        this.secondFinisher = second.finisher();
+        this.thirdFinisher = third.finisher();
+        this.fourthFinisher = fourth.finisher();
+    }
+
+    @Override
+    public Supplier<Quadruple<ResultHolder1_, ResultHolder2_, ResultHolder3_, ResultHolder4_>> supplier() {
+        return () -> Quadruple.of(firstSupplier.get(), secondSupplier.get(), thirdSupplier.get(), fourthSupplier.get());
+    }
+
+    @Override
+    public TriFunction<Quadruple<ResultHolder1_, ResultHolder2_, ResultHolder3_, ResultHolder4_>, A, B, Runnable>
+            accumulator() {
+        return (resultHolder, a, b) -> composeUndo(firstAccumulator.apply(resultHolder.getA(), a, b),
+                secondAccumulator.apply(resultHolder.getB(), a, b),
+                thirdAccumulator.apply(resultHolder.getC(), a, b),
+                fourthAccumulator.apply(resultHolder.getD(), a, b));
+    }
+
+    private static Runnable composeUndo(Runnable first, Runnable second, Runnable third,
+            Runnable fourth) {
+        return () -> {
+            first.run();
+            second.run();
+            third.run();
+            fourth.run();
+        };
+    }
+
+    @Override
+    public Function<Quadruple<ResultHolder1_, ResultHolder2_, ResultHolder3_, ResultHolder4_>, Result_> finisher() {
+        return resultHolder -> composeFunction.apply(firstFinisher.apply(resultHolder.getA()),
+                secondFinisher.apply(resultHolder.getB()),
+                thirdFinisher.apply(resultHolder.getC()),
+                fourthFinisher.apply(resultHolder.getD()));
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object)
+            return true;
+        if (object == null || getClass() != object.getClass())
+            return false;
+        ComposeFourBiCollector<?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?> that =
```

Comment:
I'd argue we can avoid this craziness with `var`.
The type is already on the right-hand side of the assignment, no need to duplicate it.

Let's try to apply it consistently.